docs/architecture/adr-006-light-client-concurrency.md

```diff
@@ -0,0 +1,106 @@
+# ADR 006: Light Client Concurrency
+
+## Changelog
+- 2020-03-15: Initial draft.
+
+## Context
+
+The light client protocol provides a method for verifying application
+state without the execution of all preceding transactions. The
+protocol for the light client is described in
+[English](https://github.com/tendermint/spec/tree/bucky/light-reorg/spec/consensus/light)
+while the rust implementation is described in
+[ADR-002](adr-002-light-client-adr-index.md). This ADR outlines the
+next iteration of the light client implementation in rust in which
+multiple modules (Peer Exchange, Requester, Verifier, etc) operate
+concurrently.
+
+The basis for this work comes from the learnings of the [Reactor
+Experiments](https://github.com/informalsystems/reactor-experiments) as
+well as the [Blockchain Reactor
+Refactor](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-043-blockchain-riri-org.md). The key take aways from that work are:
+
+1. Communicate with events and not state
+2. Separate concerns into independently verifiable (by humans as well as
+   computers) components.
+3. Push IO to the edge of the system to allow simulation of components
+   as well as integration between components.
+
+## Decision
+
+Model the Peer Exchange, Requester, Detector and Verifier as separate
+independently verifiable components. Each component is a finite state
+machine with internal state. Communication between components is done by
+event dispatching to a routing layer. The routing layer handles input
+and output by routing to/from peers to components and between
+components. The routing layer will provide the necessary abstraction to
+allow simulation of individual components as well as integration between
+components and peers.
+
+![Concurrency Diagram](assets/light-client-concurrency.png)
+
+This architecture outlines components that have been 
+described in previous ADRs and will only be expanded upon here.
+
+### Peer Exchange
+The light clients requires an expanding set of peers to get headers
+from. The larger and more diverse the set of peers; the better the
+chances that the light client will detect a fork. The peer list can be
+seeded with peers from configuration and will can crawl the network.
```

Comment:
Here is a question. 

Should the light client be running the light client protocol only with 1 peer and only running some kind minimal fork detection protocol with other peers?

I don't think it makes any sense to be running the entire light client protocol on multiple peers.

Reply:
By running the entire light client protocol on multiple peers do you mean fetching headers from multiple peers? I think the idea is to get the primary header from a single peer and then asynchronously fetch the same header from multiple peers to perform fork detection. If the headers differ, then the different header will also go through verification. If the header differs and also pass verification, we have a fork on the main chain. Is that what you mean?

Reply:
Yes this is what I'm thinking.